### DISCUSSION TITLE
do Socket.Receive/SendAsync(Memory<byte>) allocate memory on each call?

### DISCUSSION THREAD

**Discussion by freddyrios** [dotnet/runtime#104475] [General] [answered]
Specifically I am referring to allocating new AwaitableSocketAsyncEventArgs.

When I read this in the Socket.Tasks file it looked like there would be some reuse: https://github.com/dotnet/runtime/blob/8bcbe18072866a367d2041fe521c94766a4a9401/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs#L18-L19

However, It seems **every use sets it to null** and do nothing with the previous value (as its null).

This pattern in ConnectAsync is the same in multiple places, including ReceiveAsync. Sets it to null and if the previous value was null creates a new instance (but does not set the field):
https://github.com/dotnet/runtime/blob/8bcbe18072866a367d2041fe521c94766a4a9401/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs#L93-L96

Sets it to null and if the previous was not null dispose it:
https://github.com/dotnet/runtime/blob/8bcbe18072866a367d2041fe521c94766a4a9401/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs#L956

Sets it to null if the previous value matched the current instance + disposes the current instance:
https://github.com/dotnet/runtime/blob/8bcbe18072866a367d2041fe521c94766a4a9401/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Tasks.cs#L1032-L1036

Copies the existing value of the field to another instance:
https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Sockets/src/System/Net/Sockets/Socket.Unix.cs#L281

**Comment by elgonzo** [comment, marked as answer]
> Sets it to null if the previous value matched the current instance + disposes the current instance:

You did read the `Interlocked.CompareExchange` wrong. The last parameter in `Interlocked.CompareExchange(ref cache, this, null)` is the value compared against. Therefore, it sets _cache_ to _this_ if the value of _cache_ is _null_ (and disposes the current instance if there is already some non-null instance in _cache_).

**Reply by freddyrios** [reply]
Thanks, that makes more sense, sry for the noise 😅
